core/src/main/scala/org/http4s/client/jdkhttpclient/WebSocketClient.scala

```diff
@@ -0,0 +1,167 @@
+package org.http4s.client.jdkhttpclient
+
+import cats._
+import cats.data.Chain
+import cats.effect._
+import cats.effect.concurrent.{Deferred, Ref, TryableDeferred}
+import cats.implicits._
+import fs2.{Pipe, Pull, Stream}
+import org.http4s.{Headers, Method, Uri}
+import scodec.bits.ByteVector
+
+/**
+  * A websocket request.
+  *
+  * @param uri The URI.
+  * @param headers The headers to send. Some websocket clients reject protocol-specific headers.
+  * @param method The method of the intial HTTP request. Ignored by some clients.
+  * @param subprotocols The subprotocols, descending from most prefered.
+  */
+case class WSRequest(
+    uri: Uri,
+    headers: Headers = Headers.empty,
+    method: Method = Method.GET,
+    subprotocols: List[String] = List.empty
+)
+
+sealed trait WSFrame
+
+sealed trait WSControlFrame extends WSFrame
+
+sealed trait WSDataFrame extends WSFrame
+
+object WSFrame {
+  final case class Close(statusCode: Int, reason: String) extends WSControlFrame
+  final case class Ping(data: ByteVector) extends WSControlFrame
+  final case class Pong(data: ByteVector) extends WSControlFrame
+  final case class Text(data: String, last: Boolean = true) extends WSDataFrame
+  final case class Binary(data: ByteVector, last: Boolean = true) extends WSDataFrame
+}
+
+trait WSConnection[F[_]] {
+
+  /** Send a single websocket frame. The sending side of this connection has to be open. */
+  def send(wsf: WSFrame): F[Unit]
+
+  /** Send multiple websocket frames. Equivalent to multiple `send` calls, but at least as fast. */
+  def sendMany[G[_]: Traverse, A <: WSFrame](wsfs: G[A]): F[Unit]
+
+  /** A `Pipe` which sends websocket frames and emits `()`'s when they are sent. */
+  final def sendPipe: Pipe[F, WSFrame, Unit] = _.chunks.evalMap(sendMany(_))
+
+  /** Wait for a single websocket frame to be received. Returns `None` if the receiving side is closed.  */
+  def receive: F[Option[WSFrame]]
+
+  /** A stream of the incoming websocket frames. */
+  final def receiveStream: Stream[F, WSFrame] = Stream.repeatEval(receive).unNoneTerminate
+
+  /** The negotiated subprotocol, if any. */
+  def subprotocol: Option[String]
+
+}
+
+trait WSConnectionHighLevel[F[_]] {
+
+  /** Send a single websocket frame. The sending side of this connection has to be open. */
+  def send(wsf: WSDataFrame): F[Unit]
+
+  /** Send multiple websocket frames. Equivalent to multiple `send` calls, but at least as fast. */
+  def sendMany[G[_]: Traverse, A <: WSDataFrame](wsfs: G[A]): F[Unit]
+  final def sendPipe: Pipe[F, WSDataFrame, Unit] = _.chunks.evalMap(sendMany(_))
+
+  /** Send a Ping frame. */
+  def sendPing(data: ByteVector = ByteVector.empty): F[Unit]
```

Comment:
Does this (and `sendClose`) have a reasonable default implementation?

Reply:
Not in terms of another method, as `send` can only send data frames.

Reply:
Ah, right.  I was thinking solely in terms of `WSFrame`, and not the intermediate types.